Add the `decoratorsAutoAccessors` parser plugin

pzuraq opened the pull request:
| Q                        | A <!--(Can use an emoji 👍) -->
| ------------------------ | ---
| Fixed Issues?            |  N/A <!-- remove the (`) quotes and write "Fixes" before the number to link the issues -->
| Patch: Bug Fix?          | No
| Major: Breaking Change?  | No
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR Link    | N/A<!-- If only readme change, add `[skip ci]` to your commits -->
| Any Dependency Changes?  | No
| License                  | MIT

This is the first part of the implementation of the latest version of
the decorators proposal: https://github.com/tc39/proposal-decorators

The keyword is outlined here in the README: https://github.com/tc39/proposal-decorators#class-auto-accessors

It adds a `simpleAutoAccessors` plugin which can be used to parse the
`accessor` keyword into ClassAccessorProperty and ClassPrivateAccessorProperty
node types.

<!--
Before making a PR, please read our contributing guidelines
https://github.com/babel/babel/blob/main/CONTRIBUTING.md

Please note that the Babel Team requires two approvals before merging most PRs.

For issue references: Add a comma-separated list of a [closing word](https://help.github.com/articles/closing-issues-via-commit-messages/) followed by the ticket number fixed by the PR. (it should be underlined in the preview if done correctly)

If you are making a change that should have a docs update: submit another PR to https://github.com/babel/website
-->



<!-- Describe your changes below in as much detail as possible -->

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/13681"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

fedeci commented on packages/babel-parser/src/parser/statement.js:
```diff
@@ -1563,6 +1567,24 @@ export default class StatementParser extends ExpressionParser {
 
       this.checkGetterSetterParams(publicMethod);
+    } else if (
+      isContextual &&
+      key.name === "accessor" &&
+      !this.isLineTerminator()
```
Probably this branch may also be merged with the following one.
```suggestion
      isContextual &&
      key.name === "accessor" &&
      this.hasPlugin("decorators") &&
      !this.isLineTerminator()
```

fedeci replied:
Same as https://github.com/babel/babel/pull/13681#discussion_r689425353

pzuraq replied:
I tried to merge the two branches, but it results in some nested `if` logic that actually gets fairly confusing. If you want me to do it I can still, but I think it's much clearer this way (if a little more verbose) so decided to keep it.

fedeci replied:
Just keep it as is then.

fedeci commented on packages/babel-parser/test/fixtures/experimental/accessor/basic-accessor/input.js:
```diff
@@ -0,0 +1,3 @@
+class Foo {
```
These tests need to be moved under `decorators/` folder.

fedeci commented:
`Spec compliance` or `New feature`?

nicolo-ribaudo commented:
New feature, it's the first step to implement the new proposal.

nicolo-ribaudo commented on packages/babel-parser/src/parser/statement.js:
```diff
@@ -1373,6 +1373,7 @@ export default class StatementParser extends ExpressionParser {
       prop.computed = false;
       prop.key = key;
       prop.static = false;
+      prop.accessor = false;
```
This should be more like
```js
if (this.hasPlugins("decorators") && this.getPluginOption("decorators", "version") === "X")
```
where "X" is a name we give to the current version (it could even just be the date of the last meeting where the current proposal was presented).

JLHwung commented:
On the AST changes, can you open a PR to the [ESTree](https://github.com/estree/estree) project? See also https://github.com/babel/babel/blob/main/CONTRIBUTING.md#creating-a-new-plugin-spec-new for the workflow of creating a new plugin.

pzuraq commented on packages/babel-parser/package.json:
```diff
@@ -4,7 +4,7 @@
   "author": "The Babel Team (https://babel.dev/team)",
   "homepage": "https://babel.dev/docs/en/next/babel-parser",
-  "bugs": "https://github.com/babel/babel/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3A%22pkg%3A+parser+%28babylon%29%22+is%3Aopen",
+  "bugs": "https://github.com/babel/babel/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3A%22pkg%3A+parser+%28bab;lon%29%22+is%3Aopen",
```
huh, don't know how this snuck in

nicolo-ribaudo commented:
We don't follow ESTree for class fields anyway, right?

pzuraq commented on packages/babel-parser/test/fixtures/experimental/decorators-modern/options.json:
```diff
@@ -0,0 +1,3 @@
+{
+  "plugins": [["decorators", { "decoratorsBeforeExport": false, "version": "modern" }]]
```
We haven't discussed this in a long time, but I believe the status quo is that decorators will come after exports. There are some objectors to this, notably the Angular team, but most other objectors have withdrawn at this point.

nicolo-ribaudo commented:
I'm reading https://github.com/tc39/proposal-decorators/issues/426, https://github.com/tc39/proposal-decorators#class-auto-accessors and https://github.com/tc39/proposal-grouped-and-auto-accessors, and I find it slightly annoying that we have to implement this feature under the `"decorators"` plugin.

It _feels_ like it should go under the plugin for https://github.com/tc39/proposal-grouped-and-auto-accessors. I 100% understand that this feature is needed for the decorators MVP, but it feels like it needs to be there just because the decorators proposal happens to be Stage 2 while https://github.com/tc39/proposal-grouped-and-auto-accessors is Stage 1.

There are a few options, considering a feature where we'll implement https://github.com/tc39/proposal-grouped-and-auto-accessors:
1. `accessor x = 1` is enabled both by the `decorators` and the `autoAccessors` plugin
2. `accessor x = 1` is only enabled by `autoAccessors`. We now add an incomplete `autoAccessors` plugin just for what's needed by the decorators proposal, and we'll complete it in the future.
3. `accessor x = 1` is enabled by `autoAccessors` and by a temporary `simpleAutoAccessors` plugin (which is the intersection between the two proposals). When we'll release `autoAccessors`, we will documentation-deprecate `simpleAutoAccessors`.

I slightly prefer 3, but I'd like to wait for the opinion of other team members. The same question applies to the Babel transform plugin.

Also, regardless of what we choose, we should design the AST for `accessor x = 1` to be forward-compatible with the other accessors proposal. I think I initially said that the `.accessor: boolean` property was ok, but now that I'm more aware about the grouped accessors proposal I fear that it won't be compatible with things like `accessor x { get } = 1`.

nicolo-ribaudo commented:
Or maybe it's ok to use `.accessor: true`, and we'll introduce `ClassAccessorProperty` only when the accessor has explicitly `{ ... }`.

pzuraq commented:
@nicolo-ribaudo that's what I was thinking, since grouped and auto-accessors is currently at stage 1 and there's no guarantee that it will advance, there are a lot of concerns with certain aspects of the proposal. It is a bit annoying though, I would be open to adding a new type of node for this if you prefer to keep a cleaner separation.

I think option 3 makes the most sense as well, I can update the implementation accordingly.

JLHwung commented:
> We don't follow ESTree for class fields anyway, right?

Yes, the class fields unfortunately differ between Babel and ESTree, but we could coordinate early, for example to add `accessor: boolean` to both Babel's `ClassProperty` and ESTree's equivalent `PropertyDefinition` so we don't end up having more differences, let's say, `hasAccessor: boolean` in `PropertyDefinition` but not the other.

The idea is to port early proposals to the ESTree so in the future we don't introduce more discrepancy between two AST. E.g. `StaticBlock` is identical in Babel and ESTree.

JLHwung commented on packages/babel-parser/test/fixtures/experimental/decorators-modern/accessor/basic-accessor/input.js:
```diff
@@ -0,0 +1,3 @@
+class Foo {
+  accessor bar;
```
Can you add a test case for static accessor?
```js
class Foo {
  static accessor bar;
}
```

pzuraq commented:
Alright, this PR has been updated to add the `simpleAutoAccessors` plugin that @nicolo-ribaudo outlined. I also added a number of test cases to cover the different combinations of modifiers and options to accessor properties. 

@JLHwung I'll make a separate PR to add the types to ESTree as soon as I have time, thanks for pointing that out!

fedeci commented on packages/babel-parser/test/fixtures/experimental/decorators-modern/options.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "plugins": [
+    ["decorators", { "decoratorsBeforeExport": false, "version": "modern" }],
```
This should be no longer needed

nicolo-ribaudo commented on packages/babel-parser/src/parser/statement.js:
```diff
@@ -1569,6 +1573,23 @@ export default class StatementParser extends ExpressionParser {
+    ) {
+      this.resetPreviousNodeTrailingComments(key);
+
+      // The so-called parsed name would have been "get/set": get the real name.
```
```suggestion
      // The so-called parsed name would have been "accessor": get the real name.
```

nicolo-ribaudo reviewed (commented):
Some missing tests with newlines:
- ```js
  class A {
    accessor
    x
  }
  ```
  
- ```js
  class A {
    static accessor
    x
  }
  ```

- ```js
  class A {
    accessor
    [x]
  }
  ```

And a test for invalid swapped static/accessor:

- ```js
  class A {
    accessor static x
  }
  ```
- ```js
  class A {
    accessor static
    x
  }
  ```

JLHwung commented:
@pzuraq I have open an AST PR for ESTree: https://github.com/estree/estree/pull/255 feedbacks are welcome.

nicolo-ribaudo commented:
@JLHwung We diverge from ESTree when representing private fields and methods. I think we should also diverge when representing private accessors, using `ClassPrivateAccessor`, for internal consistency.

JLHwung commented:
I tend to merge `PrivateClassAccessor` into `ClassAccessor` because unlike class methods/properties, a class accessor can have a public getter and a private setter:

```js
class C { accessor a { get; #set } }
```

Should it be a `PrivateClassAccessor` or a `PublicClassAccessor`? I think people may have different opinions here but we don't have to classify them if we cover them with `ClassAccessor`.

nicolo-ribaudo commented:
Ok, fair. I was thinking about the `accessor #x = y` case, but it doesn't map well to the extended proposal.

pzuraq commented:
Ok, this PR has been updated with the extra test cases you gave @nicolo-ribaudo, and I've fixed the failing tests. Couple notes:

- In keeping with the ESTree spec proposal, I changed `init` back to `value` for the accessor nodes, so they can essentially be extensions of class property/class private property. If we still want to have it be init instead, I can change it back.
- For some reason, `computed` shows up on `ClassPrivateAccessorProperty`. I can't figure out how to remove it, and it isn't really necessary. Happy to remove if anyone can figure out how to do that.

Other than those issues, I think this PR is good to go!

JLHwung reviewed (commented):
> For some reason, computed shows up on ClassPrivateAccessorProperty

Good catch! The `computed: false` is assigned by `parsePropertyName` when we parse `accessor` as a property name. Since then we decide to reinterpret it as an accessor keyword, the `computed` property gets no chance to be deleted.

Currently we have similar issues on private accessors / async private methods. The AST will contain `computed: false` even though `computed` is not in the `ClassPrivateMethod` interface.

Note that deleting a property from AST node is a perf killer. To fix this issue we can move `computed` assignment out of `parsePropertyName` and assign it when we are sure the consumed token is a class modifier / key identifier. This can be in a separate PR so it can get merged in patch releases.

pzuraq commented on packages/babel-compat-data/data/plugins.json:
```diff
@@ -379,7 +379,7 @@
     "opera": "33",
     "edge": "14",
-    "firefox": "45",
+    "firefox": "41",
```
I believe lint was failing and it told me to run that command, but maybe it was only a local issue. I can revert this and see if it passes

JLHwung replied:
You can also try rebasing this PR on upstream main and re-run `yarn`. I guess compat-data is out of dated.

pzuraq commented on packages/babel-types/src/definitions/core.ts:
```diff
@@ -2117,6 +2117,55 @@ defineType("ClassProperty", {
+      optional: true,
+    },
+    typeAnnotation: {
+      validate: process.env.BABEL_8_BREAKING
```
How do we do that exactly? Not familiar with the breaking behavior, I copied this from the other definitions

pzuraq commented on packages/babel-types/src/definitions/core.ts:
```diff
@@ -2157,6 +2206,49 @@ defineType("ClassPrivateProperty", {
   },
 });
 
+defineType("ClassPrivateAccessorProperty", {
```
I think that if we do not, we probably need to add some level of validation. There are a lot of extra properties, such as `computed`, on the public property which do not apply to the private one. 

I think my biggest worry is that this will result in a very different API than the norms that were established by `ClassPrivateProperty` and `ClassPrivateMethod`. Users of plugins will likely expect other class elements to follow that convention, and there will probably be some amount of code that has to be somewhat different in order to support it.

I guess I think it would make the most sense if all of these were valid:

```js
class Foo {
  accessor a { get; set; }
  accessor #a { get; set; }
  accessor a { #get; #set; }
  accessor #a { #get; #set; }
}
```

But I think the last case there is invalid, if it's already private then it can't be made _more_ private, so I think the distinction can still be useful and is keeping with the convention.

JLHwung replied:
> accessor a { #get; #set; }

`#get` is not proposed so `#get` is invalid.

> accessor #a { #get; #set; }

True. This is invalid because only accessor properties with Identifier names can have a #set stub or #set method.

The case `accessor a { get; #set; }` pushes me to merge the two, because we still need to handle the private elements in an `ClassAccessorProperty`. For example we can think of the transformer routine:

```js
if (path.isClassAccessorProperty()) {
  if (path.get("set").isPrivate()) {
    // handle private
  }
  // handle public
} else if (path.isClassPrivateAccessorProperty()) {
  // handle private
}
```
versus
```js
if (path.isClassAccessorProperty()) {
  if (path.get("key").isPrivate || path.get("set").isPrivate()) {
    // handle private
  }
  // handle public
}
```

pzuraq replied:
I think it's pretty likely that the transforms for `accessor a { get; #set; }` will be fairly different from `accessor #a { get; set; }`, so I'm not sure if it will end up being that much of a benefit, but if it makes more sense to represent them with the same node type then we can do that.

nicolo-ribaudo commented:
We (@pzuraq, @JLHwung and me) just had a call to discuss about this PR.
- The plugin name should be `decoratorsAutoAccessors`, to make it clear that it's only meant to be used for the intersection between the "class accessors" and "decorators" proposals.
- Chris will try following the ESTree AST shape, but if it ends up being too complex we will fallback to a simple `.accessor: boolean` on `ClassProperty` and `ClassPrivateProperty`. Regardless of what path we choose, when implementing the full class accessors plugin we can change the AST in a minor since it would be opt-in.

nicolo-ribaudo commented on packages/babel-parser/src/parser/statement.js:
```diff
@@ -1585,6 +1587,18 @@ export default class StatementParser extends ExpressionParser {
+    } else if (
+      isContextual &&
+      key.name === "accessor" &&
+      this.hasPlugin("simpleAutoAccessors") &&
```
This check can be moved to _inside_ the `if`, using `expectPlugin` rather than `hasPlugin` to report better errors ("missing plugin" vs "missing semicolon" probably)

JLHwung replied:
Please also add a new test case in `packages/babel-parser/test/fixtures/experimental/_no-plugin`.

nicolo-ribaudo commented on packages/babel-parser/src/parser/statement.js:
```diff
@@ -1666,6 +1680,33 @@ export default class StatementParser extends ExpressionParser {
+  ) {
+    if (
+      !prop.computed &&
+      prop.key.type !== "PrivateName" &&
```
When the `["estree", { "classFeatures": true }]` plugin is enabled, the type of private names is diferent. I think we have a `this.isPrivateName(prop.key)` (or something similar) method to handle both the AST shapes.

JLHwung replied:
```suggestion
      !isPrivate &&
```

pzuraq replied:
I don't think that `isPrivate` does the type guarding correctly, which is the issue here, so going to go with `this.isPrivateName` to get that assertion

pzuraq replied:
actually it is more performant not to do the check probably, so I can just use type casting

JLHwung replied:
Yeah you can perfectly ignore the FlowError `// $FlowIgnore` or use `/*:: invariant(prop.key.type !== "PrivateName") */`. Flow does not understand the equivalence between the token type and AST node type anyway.

JLHwung commented on packages/babel-parser/src/types.js:
```diff
@@ -849,6 +850,20 @@ export type ClassPrivateProperty = NodeBase & {
+  DeclarationBase & {
+    type: "ClassAccessorProperty",
+    key: Expression | PrivateName,
+    value: ?Expression, // TODO: Not in spec that this is nullable.
```
Nit: This todo item can be removed. It is extremely out-of-dated as there is `PropertyDefinition` in estree with nullable `value`: https://github.com/estree/estree/blob/master/es2022.md#propertydefinition

nicolo-ribaudo reviewed (approved):
Thanks!

JLHwung commented on packages/babel-generator/src/generators/classes.ts:
```diff
@@ -114,6 +114,48 @@ export function ClassProperty(this: Printer, node: t.ClassProperty) {
+
+  this.tsPrintClassMemberModifiers(node, /* isField */ true);
+
+  this.word("accessor");
```
Nit: we should also print inner comments here, recall that in `parseClassMemberWithIsStatic` we have reset the trailing comments of `accessor` keyword to be the inner comments of accessor property:

```suggestion
  this.word("accessor");
  this.printInnerComments(node);
```

```js
class C {
  accessor /* inner comment */ [computed] = 1
}
```

JLHwung reviewed (approved):
Awesome work. The unresolved comments are nits / edge cases.

nicolo-ribaudo commented:
Thanks! Merged to https://github.com/babel/babel/tree/feat-7.16.0/decorators
